openedx/core/djangoapps/course_roles/tests/test_api.py

```diff
@@ -0,0 +1,135 @@
+        self.role_5.permissions.add(permission_5)
+
+    def test_get_all_user_permissions_for_a_course_with_anonymus_user(self):
+        """
+        Test that get_all_user_permissions_for_a_course returns an empty set when the user is anonymous
+        """
+        assert not get_all_user_permissions_for_a_course(self.anonymous_user, self.course_1_key)
+
+    def test_get_all_user_permissions_for_a_course(self):
+        """
+        Test that get_all_user_permissions_for_a_course returns the correct permissions for the user and course
+        """
+        UserRole.objects.create(
+            user=self.user_1, role=self.role_1, course_id=self.course_1.id, org=self.organization_1
+        )
+        UserRole.objects.create(
+            user=self.user_2, role=self.role_4, course_id=self.course_1.id, org=self.organization_1
+        )
+        # Test that the correct permissions are returned for user_1
+        assert get_all_user_permissions_for_a_course(self.user_1, self.course_1_key) == {self.permission_1}
+        UserRole.objects.create(
+            user=self.user_1, role=self.role_2, course_id=self.course_1.id, org=self.organization_1
+        )
+        self.clear_caches()
+        # Test that the correct permissions are returned for user_1
+        assert get_all_user_permissions_for_a_course(self.user_1, self.course_1_key) == {
+            self.permission_1, self.permission_2}
+
+        UserRole.objects.create(
+            user=self.user_1, role=self.role_3, org=self.organization_1
+        )
+        self.clear_caches()
+        # Test that the correct permissions are returned for user_1, including org level permissions
+        assert get_all_user_permissions_for_a_course(self.user_1, self.course_1_key) == {
+            self.permission_1, self.permission_2, self.permission_3}
+        UserRole.objects.create(
+            user=self.user_1, role=self.role_5
+        )
+        self.clear_caches()
+        # Test that the correct permissions are returned for user_1, including instance level permissions
+        assert get_all_user_permissions_for_a_course(self.user_1, self.course_1_key) == {
+            self.permission_1, self.permission_2, self.permission_3, self.permission_5}
+
+    def test_get_all_user_permissions_for_a_course_with_no_permissions(self):
+        """
+        Test that get_all_user_permissions_for_a_course returns an empty list when the user has no permissions
+        """
+        assert not get_all_user_permissions_for_a_course(self.user_1, self.course_1_key)
+
+    @ddt.data(
+        (None, 1),
+        (1, None),
+        (None, None),
+    )
+    @ddt.unpack
+    def test_get_all_user_permissions_for_a_course_with_none_values(self, user_id, course_key):
+        """
+        Test that get_all_user_permissions_for_a_course raises value error when the user has no permissions
+        """
+        with pytest.raises(TypeError):
+            get_all_user_permissions_for_a_course(user_id, course_key)
```

Comment:
Why is this the expected behavior, as opposed to returning an empty set to show they have no permissions for it?